src/hotspot/share/runtime/thread.cpp

```diff
@@ -435,30 +435,38 @@ void Thread::check_for_dangling_thread_pointer(Thread *thread) {
 }
 #endif
 
-// Is the target JavaThread protected by the calling Thread
-// or by some other mechanism:
-bool Thread::is_JavaThread_protected(const JavaThread* p) {
-  // Do the simplest check first:
-  if (SafepointSynchronize::is_at_safepoint()) {
-    // The target is protected since JavaThreads cannot exit
-    // while we're at a safepoint.
-    return true;
-  }
+// Is the target JavaThread protected by the calling Thread or by some other
+// mechanism? If checkTLHOnly is true (default is false), then we only check
+// if the target JavaThread is protected by a ThreadsList (if any) associated
+// with the calling Thread.
+//
+bool Thread::is_JavaThread_protected(const JavaThread* p, bool checkTLHOnly) {
+  Thread* current_thread = nullptr;
+  if (checkTLHOnly) {
+    current_thread = Thread::current();
```

Comment:
This seems redundant due to line 463. You can just have a `if (!checkTLHOnly)` block here.

Reply:
Sorry I missed that line 463 is still within the else from line 447.

Thread::current() is a compiler-defined thread-local access so should be relatively cheap these days, but I have no numbers.

Reply:
I'm really tempted to go ahead and change it to always set
current thread when it is declared and then clean things up a bit.

Reply:
Yes, ^ that might make the logic easier to follow. I can't figure out what checkTLSOnly means.  Could it be refactored into a different function like check_TLS() and then call it in the place where you pass true instead of is_JavaThread_protected?  Does checkTLSOnly skip a lot of these checks?

Reply:
I've changed `is_JavaThread_protected()` to materialize current_thread when it is declared
to simplify the code paths.

Reply:
> I can't figure out what checkTLSOnly means.

See the function's header comment:

```
// ... If checkTLHOnly is true (default is false), then we only check
// if the target JavaThread is protected by a ThreadsList (if any) associated
// with the calling Thread.
```

> Could it be refactored into a different function like check_TLS() and then
> call it in the place where you pass true instead of is_JavaThread_protected?

I had a copy of this logic in a separate function and @dholmes-ora suggested
that I add the `checkTLHOnly` parameter to this function and get rid of the
separate copy so I've been there and done that.

I could *move* the logic into a separate function and then call that new function
from here and from the other call sites that pass `true`, but then I'd have to
materialize current_thread in that new function and then on this code path we
would be materializing current_thread twice. I don't want to do that.

> Does checkTLSOnly skip a lot of these checks?

Please read the header comment to see what `checkTLHOnly` does.